execution/gethexec/caff_node.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		log.Error("failed to get state at last block header", "err", err)
+		return false
+	}
+	log.Info("Initial State", "lastBlockHash", lastBlockHeader.Hash(), "lastBlockStateRoot", lastBlockHeader.Root)
+	startTime := time.Now()
+
+	// Run the Produce block function in replay mode
+	// This is the core function that is used by replay.wasm to validate the block
+	block, receipts, err := arbos.ProduceBlock(messageWithMetadata.Message, messageWithMetadata.DelayedMessagesRead, lastBlockHeader, statedb, n.executionEngine.bc, n.executionEngine.bc.Config(), false, core.MessageReplayMode)
+	if err != nil {
+		log.Error("Failed to produce block", "err", err)
+		return false
+	}
+
+	// If block is nil or receipts is empty, return false
+	if len(receipts) == 0 || block == nil {
+		log.Error("Failed to produce block, no receipts or block")
+		return false
+	}
+	blockCalcTime := time.Since(startTime)
+
+	log.Info("Produced block", "block", block.Hash(), "blockNumber", block.Number(), "receipts", len(receipts))
+
+	err = n.executionEngine.appendBlock(block, statedb, receipts, blockCalcTime)
+	if err != nil {
+		log.Error("Failed to append block", "err", err)
+		return false
+	}
+
+	// Pop the message from the front of the queue at the end.
+	n.messagesWithMetadata = n.messagesWithMetadata[1:]
+	n.messagesWithMetadataPos = n.messagesWithMetadataPos[1:]
+
+	return true
+}
+
+func (n *CaffNode) queueMessagesFromHotshot(ctx context.Context) error {
+	if n.nextSeqBlockNum == 0 {
+		latestBlock, err := n.espressoClient.FetchLatestBlockHeight(ctx)
+		if err != nil {
+			log.Warn("unable to fetch latest hotshot block", "err", err)
+			return err
+		}
+		log.Info("Started node at the latest hotshot block", "block number", latestBlock)
+		n.nextSeqBlockNum = latestBlock
+	}
+
+	nextSeqBlockNum := n.nextSeqBlockNum
+	header, err := n.espressoClient.FetchHeaderByHeight(ctx, nextSeqBlockNum)
+	if err != nil {
+		log.Warn("failed to fetch header", "err", err)
+		return err
+	}
+	height := header.Header.GetBlockHeight()
+	arbTxns, err := n.espressoClient.FetchTransactionsInBlock(ctx, height, n.namespace)
+	if err != nil {
+		log.Warn("failed to fetch transactions", "err", err)
+		return err
+	}
+	if len(arbTxns.Transactions) == 0 {
+		return nil
+	}
+
+	n.messagesStateMutex.Lock()
+	defer n.messagesStateMutex.Unlock()
+	for _, tx := range arbTxns.Transactions {
+		// Parse hotshot payload
+		_, indices, messages, err := arbutil.ParseHotShotPayload(tx)
+		if err != nil {
+			log.Warn("failed to parse hotshot payload, will retry", "err", err)
+			return err
+		}
+		// Parse the messages
+		for i, message := range messages {
+			var messageWithMetadata arbostypes.MessageWithMetadata
+			err = rlp.DecodeBytes(message, &messageWithMetadata)
+			if err != nil {
+				log.Warn("failed to decode message, will retry", "err", err)
+				return err
+			}
+			n.messagesWithMetadata = append(n.messagesWithMetadata, &messageWithMetadata)
```

Comment:
We fetch the hotshot transactions consecutively and append them one by one. And the blocks are created according to this queue. My concern here is that hotshot transactions may have different order of L2 messages, won't this leads to a different status?
I think we should add the messages to tx streamer first(it already has the mechanism dealing with reorgs) and then fetch messages from there?

Reply:
https://github.com/EspressoSystems/nitro-espresso-integration/pull/477/commits/a2dedfba8179d5282b75e24b0302bfcc58550a31

Reply:
I fixed the duplicates and ordering issue here, can you see if this works?